### DISCUSSION TITLE
Rendering Custom Trait in Grapesjs React

### DISCUSSION THREAD

**Discussion by RakulAgn** [GrapesJS/grapesjs#6057] [Q&A]
Has anyone tried using a custom trait in a React wrapper? I'm having trouble accessing the value inside the custom trait.

The value comes from a React component, and I'm using `onChange` to set the value in the trait like this:

```javascript
onChange: (value: boolean) => trait.setValue(value),
```

If this is correct, the value should be set in the trait.

Here’s how I've defined the trait:
```
traits: [
  {
    label: "Background Color Full Width",
    type: "full-width-bg-color",
    id: "full-width-bg-color",
    name: "full-width-bg-color",
  },
],
```

I should be able to access the trait during an onEvent change, right? Here's my onEvent implementation:
```
tm.addType("full-width-bg-color", {
  onEvent({ component, trait }) {
    console.log(trait, component);
  },
});
```

Any ideas on why I might not be getting the value? expecting something like creating only the functionality of the trait inside the plugin but the ui should be react component 

@artf

**Comment by parthrc** [comment]
Hello @RakulAgn , I am also going through a similar problem, have you managed to solve this issue? I can update the traists from inside the react component but dont know how to rerender the component with the updated trait value.

**Reply by RakulAgn** [reply]
@parthrc I have not yet figured out. but its possible i think so in studio they have used some traits through react but its not custom trait with custom ui they have created with input type select type with that we can pass the data but for custom trait i think we cant do with that yet to confirm

**Reply by parthrc** [reply]
@artf 
@RakulAgn , working on it, will update if I figure out

**Reply by RakulAgn** [reply]
Yeah Sure Looking forward @parthrc

**Comment by RakulAgn** [comment]
@parthrc Hey I Figured Out to Use Custom Trait

**Reply by maxming2333** [reply]
I am using Custom Trait today. After debugging it for a day, I found that it cannot be used.

Finally, I found this issue. We have a strong demand for it. Please support it.

@artf

**Reply by maxming2333** [reply]
Can this PR solve the problem? 
https://github.com/GrapesJS/react/pull/7

**Reply by RakulAgn** [reply]
@maxming2333 Have you tried like this below we can pass the trait values from traits provider in react that values will be captured in components setvalue and getvalue


<img width="1172" alt="Screenshot 2024-10-06 at 11 01 25 AM" src="https://github.com/user-attachments/assets/04bba7e3-bb8a-45af-b446-fcdd2e018ea0">

```
function getSwapValue(component: Component) {
  try {
    const styles = component.getStyle();
    const direction = styles["direction"];
    return direction;
  } catch (error) {
    return "ltr";
  }
}

function setSwapValue(
  value: string,
  component: Component,
  callback: () => void
) {
  try {
    if (value === "rtl") {
      component.addStyle({ direction: "rtl" });
    } else if (value === "ltr") {
      component.addStyle({ direction: "ltr" });
    }

    callback();
  } catch (error) {
    console.log("ajksfn");
  }
}
```

**Reply by maxming2333** [reply]
@RakulAgn OK, I'll give it a try.
I tried this way before, but unfortunately it didn't work properly.

```js
  editor.Traits?.addType('news-id', {
    onEvent(...args) {
      console.log('xxxx-------->onEvent', ...args);
    },
    onUpdate(...args) {
      console.log('xxxx-------->onUpdate', ...args);
    },
  });
```

**Reply by RakulAgn** [reply]
No This Wont Work I tried Before Like this and Finally Figured it out.
